zcash_primitives/src/transaction/mod.rs

```diff
@@ -0,0 +1,179 @@
+                reader.read_exact(&mut joinsplit_sig)?;
+            }
+            jss
+        } else {
+            vec![]
+        };
+
+        let binding_sig =
+            match is_sapling_v4 && !(shielded_spends.is_empty() && shielded_outputs.is_empty()) {
+                true => Some(Signature::read(&mut reader)?),
+                false => None,
+            };
+
+        Ok(Transaction {
+            overwintered,
+            version,
+            version_group_id,
+            vin,
+            vout,
+            lock_time,
+            expiry_height,
+            value_balance,
+            shielded_spends,
+            shielded_outputs,
+            joinsplits,
+            joinsplit_pubkey,
+            joinsplit_sig,
+            binding_sig,
+        })
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        writer.write_u32::<LittleEndian>(self.header())?;
+        if self.overwintered {
+            writer.write_u32::<LittleEndian>(self.version_group_id)?;
+        }
+
+        let is_overwinter_v3 = self.overwintered
+            && self.version_group_id == OVERWINTER_VERSION_GROUP_ID
+            && self.version == OVERWINTER_TX_VERSION;
+        let is_sapling_v4 = self.overwintered
+            && self.version_group_id == SAPLING_VERSION_GROUP_ID
+            && self.version == SAPLING_TX_VERSION;
+        if self.overwintered && !(is_overwinter_v3 || is_sapling_v4) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Unknown transaction format",
+            ));
+        }
+
+        Vector::write(&mut writer, &self.vin, |w, e| e.write(w))?;
+        Vector::write(&mut writer, &self.vout, |w, e| e.write(w))?;
+        writer.write_u32::<LittleEndian>(self.lock_time)?;
+        if is_overwinter_v3 || is_sapling_v4 {
+            writer.write_u32::<LittleEndian>(self.expiry_height)?;
+        }
+
+        if is_sapling_v4 {
+            writer.write_i64::<LittleEndian>(self.value_balance.0)?;
+            Vector::write(&mut writer, &self.shielded_spends, |w, e| e.write(w))?;
+            Vector::write(&mut writer, &self.shielded_outputs, |w, e| e.write(w))?;
+        }
+
+        if self.version >= 2 {
+            Vector::write(&mut writer, &self.joinsplits, |w, e| e.write(w))?;
+            if !self.joinsplits.is_empty() {
+                writer.write_all(&self.joinsplit_pubkey)?;
+                writer.write_all(&self.joinsplit_sig)?;
+            }
+        }
+
+        if is_sapling_v4 && !(self.shielded_spends.is_empty() && self.shielded_outputs.is_empty()) {
+            match self.binding_sig {
+                Some(sig) => sig.write(&mut writer)?,
+                None => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::InvalidInput,
+                        "Missing binding signature",
+                    ))
+                }
+            }
+        }
```

Comment:
This returns an error if the binding signature should be present but has not been set, but it does not return an error for the converse, i.e. when the binding signature should not be present and has been set.